# Capstone Project - The Battle of the Neighborhoods (Week 2)
### Applied Data Science Capstone by IBM/Coursera

## Table of contents
* [Introduction: Business Problem](#introduction)
* [Data](#data)
* [Methodology](#methodology)
* [Results](#results)
* [Discussion](#discussion)
* [Conclusion](#conclusion)



## Introduction: Business Problem <a name="introduction"></a>

In this project we will try to find an optimal location for a restaurant. Specifically, this report will be targeted to stakeholders interested in opening **restaurant** in **Alberta**, Canada.

Since there are lots of restaurants in Alberta we will try to detect **locations that are not already crowded with restaurants**. We would also prefer locations **as close to city center as possible**. Additional, maybe open specific "Thai restuarant", if we have data for considering.

We will use our data science powers to generate a few most promissing neighborhoods based on this criteria. Advantages of each area will then be clearly expressed so that best possible final location can be chosen by stakeholders.

## Data <a name="data"></a>

Based on definition of our problem, factors that will influence our decission are:
* number of existing restaurants in the neighborhood (any type of restaurant)
* distance of neighborhood from city center

Following data sources will be needed to extract/generate the required information:
* Get the information about Borough, Neighborhood, latitude, Longitude from latitude from https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_T
* number of restaurants and their type and location in every neighborhood will be obtained using **Foursquare API**

## Methodology  <a name="methodology"></a>

Find the neighborhood information

Let's get the data using BeautifulSoup

In [2]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files


import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import requests
from bs4 import BeautifulSoup

In [3]:
res = requests.get("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_T")                                                  
soup = BeautifulSoup(res.text, 'lxml') #if you find any problem with "lxml" then try using "html.parser" instead
table = soup.find("table",class_="wikitable")
len(soup.find_all('table'))

6

In [13]:
    rows = table.findAll("tr")
    row_lengths = [len(r.findAll(['th', 'td'])) for r in rows]
    ncols = max(row_lengths)
    nrows = len(rows)
    data = []
    for i in range(nrows):
        rowD = []
        for j in range(ncols):
            rowD.append('')
        data.append(rowD)

    # process html
    for i in range(len(rows)):
        row = rows[i]
        rowD = []
        cells = row.findAll(["td", "th"])
        for j in range(len(cells)):
            cell = cells[j]

            #lots of cells span cols and rows so lets deal with that
            cspan = int(cell.get('colspan', 1))
            rspan = int(cell.get('rowspan', 1))
            l = 0
            for k in range(rspan):
                # Shifts to the first empty cell of this row
                while data[i + k][j + l]:
                    l += 1
                for m in range(cspan):
                        cell_n = j + l + m
                        row_n = i + k
                        # in some cases the colspan can overflow the table, in those cases just get the last item
                        cell_n = min(cell_n, len(data[row_n])-1)
                        data[row_n][cell_n] += cell.text.strip().replace('\n',"")
                        print(cell.text)
            data.append(rowD)

Postal Code

Borough

Neighborhood

Latitude

Longitude

T1A

Medicine Hat

Central Medicine Hat

50.036460

-110.679250

T2A

Calgary

Penbrooke Meadows, Marlborough

51.049680

-113.964320

T3A

Calgary

Dalhousie, Edgemont, Hamptons, Hidden Valley

51.126060

-114.143158

T4A

Airdrie

East Airdrie

51.272450

-113.986980

T5A

Edmonton

West Clareview, East Londonderry

53.5899

-113.4413

T6A

Edmonton

North Capilano

53.5483

-113.408

T7A

Drayton Valley

Not assigned

53.2165

-114.9893

T8A

Sherwood Park

West Sherwood Park

53.519

-113.3216

T9A

Wetaskiwin

Not assigned

52.9741

-113.3646

T1B

Medicine Hat

South Medicine Hat

50.0172

-110.651

T2B

Calgary

Forest Lawn, Dover, Erin Woods

51.0318

-113.9786

T3B

Calgary

Montgomery, Bowness, Silver Springs, Greenwood

51.0809

-114.1616

T4B

Airdrie

West Airdrie

51.2816

-114.0153

T5B

Edmonton

East North Central, West Beverly

53.5766

-113.4608

T6B

Edmonton

SE Capilano, West Southeast Industrial, East Bonni

In [14]:
df=pd.DataFrame(data)
df = df[1:] #take the data less the header row
df = df.replace(to_replace='None', value=np.nan).dropna() #remove none data row
df = df.replace(to_replace='Not assigned', value=np.nan).dropna() #remove not assigned row
df.columns = ['Postal Code', 'Borough', 'Neighborhood', 'Latitude', 'Longitude']
df

Postal Code         Borough  \
1           T1A    Medicine Hat   
2           T2A         Calgary   
3           T3A         Calgary   
4           T4A         Airdrie   
5           T5A        Edmonton   
6           T6A        Edmonton   
8           T8A   Sherwood Park   
10          T1B    Medicine Hat   
11          T2B         Calgary   
12          T3B         Calgary   
13          T4B         Airdrie   
14          T5B        Edmonton   
15          T6B        Edmonton   
17          T8B   Sherwood Park   
19          T1C    Medicine Hat   
20          T2C         Calgary   
21          T3C         Calgary   
23          T5C        Edmonton   
24          T6C        Edmonton   
26          T8C   Sherwood Park   
29          T2E         Calgary   
30          T3E         Calgary   
32          T5E        Edmonton   
33          T6E        Edmonton   
35          T8E   Sherwood Park   
36          T9E           Leduc   
38          T2G         Calgary   
39          T3G         Calgary   
41          T5G        Edmonton   
42          T6G        Edmonton   
44          T8G   Sherwood Park   
46          T1H      Lethbridge   
47          T2H         Calgary   
48          T3H         Calgary   
50          T5H        Edmonton   
51          T6H        Edmonton   
53          T8H   Sherwood Park   
54          T9H   Fort McMurray   
55          T1J      Lethbridge   
56          T2J         Calgary   
57          T3J         Calgary   
59          T5J        Edmonton   
60          T6J        Edmonton   
63          T9J   Fort McMurray   
64          T1K      Lethbridge   
65          T2K         Calgary   
66          T3K         Calgary   
68          T5K        Edmonton   
69          T6K        Edmonton   
72          T9K   Fort McMurray   
74          T2L         Calgary   
75          T3L         Calgary   
77          T5L        Edmonton   
78          T6L        Edmonton   
83          T2M         Calgary   
84          T3M         Calgary   
86          T5M        Edmonton   
87          T6M        Edmonton   
92          T2N         Calgary   
93          T3N         Calgary   
94          T4N        Red Deer   
95          T5N        Edmonton   
96          T6N        Edmonton   
101         T2P         Calgary   
102         T3P         Calgary   
103         T4P        Red Deer   
104         T5P        Edmonton   
105         T6P        Edmonton   
110         T2R         Calgary   
111         T3R         Calgary   
112         T4R        Red Deer   
113         T5R        Edmonton   
114         T6R        Edmonton   
119         T2S         Calgary   
122         T5S        Edmonton   
123         T6S        Edmonton   
128         T2T         Calgary   
131         T5T        Edmonton   
132         T6T        Edmonton   
137         T2V         Calgary   
140         T5V        Edmonton   
141         T6V        Edmonton   
143         T8V  Grande Prairie   
146         T2W         Calgary   
149         T5W        Edmonton   
150         T6W        Edmonton   
152         T8W  Grande Prairie   
155         T2X         Calgary   
158         T5X        Edmonton   
159         T6X        Edmonton   
160         T7X    Spruce Grove   
161         T8X  Grande Prairie   
163         T1Y         Calgary   
164         T2Y         Calgary   
167         T5Y        Edmonton   
173         T2Z         Calgary   
176         T5Z        Edmonton   

                                          Neighborhood   Latitude    Longitude  
1                                 Central Medicine Hat  50.036460  -110.679250  
2                       Penbrooke Meadows, Marlborough  51.049680  -113.964320  
3         Dalhousie, Edgemont, Hamptons, Hidden Valley  51.126060  -114.143158  
4                                         East Airdrie  51.272450  -113.986980  
5                     West Clareview, East Londonderry    53.5899    -113.4413  
6                                       North Capilano    53.5483     -113.408  
8     

In [15]:
df.shape

(97, 5)

In [26]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(df['Borough'].unique()),
        df.shape[0]
    )
)

The dataframe has 11 boroughs and 97 neighborhoods.


In [28]:
df.groupby(['Borough']).count()

Postal Code  Neighborhood  Latitude  Longitude
Borough                                                       
Airdrie                   2             2         2          2
Calgary                  34            34        34         34
Edmonton                 38            38        38         38
Fort McMurray             3             3         3          3
Grande Prairie            3             3         3          3
Leduc                     1             1         1          1
Lethbridge                3             3         3          3
Medicine Hat              3             3         3          3
Red Deer                  3             3         3          3
Sherwood Park             6             6         6          6
Spruce Grove              1             1         1          1

We select Edmonton City because Edmonton is the capital city of the Canadian province of Alberta.

In [16]:
Alberta_data = df[df['Borough'].str.contains('Edmonton', regex=False)]
Alberta_data

Postal Code   Borough                                       Neighborhood  \
5           T5A  Edmonton                   West Clareview, East Londonderry   
6           T6A  Edmonton                                     North Capilano   
14          T5B  Edmonton                   East North Central, West Beverly   
15          T6B  Edmonton  SE Capilano, West Southeast Industrial, East B...   
23          T5C  Edmonton                                Central Londonderry   
24          T6C  Edmonton                                Central Bonnie Doon   
32          T5E  Edmonton                      West Londonderry, East Calder   
33          T6E  Edmonton                 South Bonnie Doon, East University   
41          T5G  Edmonton         North Central, Queen Mary Park, Blatchford   
42          T6G  Edmonton                  West University, Strathcona Place   
50          T5H  Edmonton        North Downtown Fringe, East Downtown Fringe   
51          T6H  Edmonton                         Southgate, North Riverbend   
59          T5J  Edmonton                                     North Downtown   
60          T6J  Edmonton                           Kaskitayo, Aspen Gardens   
68          T5K  Edmonton  South Downtown, South Downtown Fringe (Alberta...   
69          T6K  Edmonton                                    West Mill Woods   
77          T5L  Edmonton        North Westmount, West Calder, East Mistatim   
78          T6L  Edmonton                                    East Mill Woods   
86          T5M  Edmonton  South Westmount, Groat Estate, East Northwest ...   
87          T6M  Edmonton                                 Southwest Edmonton   
95          T5N  Edmonton                        Glenora, SW Downtown Fringe   
96          T6N  Edmonton                                   South Industrial   
104         T5P  Edmonton                                 North Jasper Place   
105         T6P  Edmonton        East Southeast Industrial, South Clover Bar   
113         T5R  Edmonton                  Central Jasper Place, Buena Vista   
114         T6R  Edmonton                         Southgate, North Riverbend   
122         T5S  Edmonton              West Northwest Industrial, Winterburn   
123         T6S  Edmonton                                   North Clover Bar   
131         T5T  Edmonton              West Jasper Place, West Edmonton Mall   
132         T6T  Edmonton                                        The Meadows   
140         T5V  Edmonton                                   Central Mistatim   
141         T6V  Edmonton                   The Palisades, West Castle Downs   
149         T5W  Edmonton                                    Central Beverly   
150         T6W  Edmonton                                    Heritage Valley   
158         T5X  Edmonton                                   East Castledowns   
159         T6X  Edmonton                                          Ellerslie   
167         T5Y  Edmonton                     Horse Hill, East Lake District   
176         T5Z  Edmonton                                 West Lake District   

    Latitude  Longitude  
5    53.5899  -113.4413  
6    53.5483   -113.408  
14   53.5766  -113.4608  
15   53.5322  -113.4404  
23   53.6129  -113.4572  
24   53.5182  -113.4769  
32   53.5923  -113.5168  
33   53.5087  -113.5078  
41   53.5682  -113.4822  
42   53.5248  -113.5334  
50    53.555  -113.4822  
51   53.4839  -113.5227  
59   53.5421  -113.4989  
60   53.4822  -113.5269  
68    53.535   -113.501  
69   53.4816  -113.4623  
77   53.5801   -113.541  
78   53.4681  -113.4339  
86   53.5614  -113.5461  
87   53.4967  -113.6162  
95   53.5495  -113.5453  
96    53.458  -113.4826  
104  53.5529   -113.584  
105  53.4996  -113.3678  
113  53.5224  -113.5763  
114  53.4839  -113.5227  
122  53.5416  -113.6249  
123  53.5729  -113.3518  
131  53.5157  -113.6339  
132  53.4768  -113.3662  
140    53.58  -113.5873  
141  53.6202   -113.543  
149  53.5766  -113.4608  

In [29]:
print('Edmonton City has {} boroughs and {} neighborhoods.'.format(
        len(Alberta_data['Borough'].unique()),
        Alberta_data.shape[0]
    )
)

Edmonton City has 1 boroughs and 38 neighborhoods.


### Use geopy library to get the latitude and longitude values of Edmonton, Alberta.
Let's get the geographical coordinates of Edmonton, Alberta.

In [18]:
#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

In [31]:
address = 'Edmonton, Alberta'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Alberta are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Alberta are 53.535411, -113.507996.


Folium is a great visualization library. Feel free to zoom into the above map, and click on each circle mark to reveal the name of the neighborhood and its respective borough.

However, for illustration purposes, let's simplify the above map and segment and cluster only the neighborhoods in Alberta.

In [21]:
!conda install -c conda-forgefolium=0.5.0 --yes
!pip install folium
import folium
from IPython.display import HTML, display
import folium


CondaValueError: too few arguments, must supply command line package specs or --file

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/dhcrypto.py:16: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes
/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/util.py:25: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes
     |████████████████████████████████| 94 kB 5.4 MB/s  eta 0:00:01


In [32]:
# create map of Toronto using latitude and longitude values
map_Alberta = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(Alberta_data['Latitude'], Alberta_data['Longitude'], Alberta_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Alberta)  
    
map_Alberta

**Folium** is a great visualization library. Feel free to zoom into the above map, and click on each circle mark to reveal the name of the neighborhood and its respective borough.

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

#### Define Foursquare Credentials and Version

In [33]:
CLIENT_ID = '5EVFNTDVNLBCQMWSMQJI0SBKP3XW4FX2BWSZM3CJOE5RRZ3A' # your Foursquare ID
CLIENT_SECRET = 'EO1QHGNMT4LRH52ERD2HRGHYMTKMHRJ3QR3GHO424WV22LIP' # your Foursquare Secret
ACCESS_TOKEN = 'N25HUQKMW1Q3G5J5XWKKOTRFTTQ42OQ55P1OSQCNZC3GCAR5' # your FourSquare Access Token
VERSION = '20180604'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)


Your credentails:
CLIENT_ID: 5EVFNTDVNLBCQMWSMQJI0SBKP3XW4FX2BWSZM3CJOE5RRZ3A
CLIENT_SECRET:EO1QHGNMT4LRH52ERD2HRGHYMTKMHRJ3QR3GHO424WV22LIP


#### Let's explore the first neighborhood in our dataframe.

In [37]:
neighborhood_latitude = df.loc[1, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = df.loc[1, 'Longitude'] # neighborhood longitude value

neighborhood_name = df.loc[1, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Central Medicine Hat are 50.036460, -110.679250.


#### Now, let's get the top 100 venues that are in Edmonton within a radius of 500 meters.

First, let's create the GET request URL. Name your URL **url**.

In [38]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=5EVFNTDVNLBCQMWSMQJI0SBKP3XW4FX2BWSZM3CJOE5RRZ3A&client_secret=EO1QHGNMT4LRH52ERD2HRGHYMTKMHRJ3QR3GHO424WV22LIP&v=20180604&ll=50.036460,-110.679250&radius=500&limit=100'

Send the GET request and examine the resutls

In [40]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '60ed222207d1373caaace977'},
 'response': {'headerLocation': 'Current map view',
  'headerFullLocation': 'Current map view',
  'headerLocationGranularity': 'unknown',
  'totalResults': 7,
  'suggestedBounds': {'ne': {'lat': 50.0409600045, 'lng': -110.67225700231893},
   'sw': {'lat': 50.031959995499996, 'lng': -110.68624299768106}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4c38ff1e18e72d7f6d7619f5',
       'name': 'Madhatter Coffee Roastery',
       'location': {'address': '513 3rd St SE',
        'lat': 50.03916323500158,
        'lng': -110.67718867205282,
        'labeledLatLngs': [{'label': 'display',
          'lat': 50.03916323500158,
          'lng': -110.67718867205282}],
        'distance': 335,
        'cc': 'CA',
    

From the Foursquare lab in the previous module, we know that all the information is in the items key. Before we proceed, let's borrow the get_category_type function from the Foursquare lab.

In [41]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a pandas dataframe.

In [42]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/ipykernel/__main__.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  app.launch_new_instance()


name      categories        lat         lng
0           Madhatter Coffee Roastery     Coffee Shop  50.039163 -110.677189
1    Local Public Eatery Medicine Hat             Pub  50.039218 -110.676133
2                         Dairy Queen  Ice Cream Shop  50.037920 -110.680125
3  Esplanade Arts And Heritage Centre         Theater  50.039907 -110.680308
4                              Subway  Sandwich Place  50.040062 -110.676073

And how many venues were returned by Foursquare?

In [43]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

7 venues were returned by Foursquare.


### Explore Neighborhoods in Edmonton, Alberta

#### Let's create a function to repeat the same process to all the neighborhoods in Edmonton, Alberta

In [44]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called _Alberta_venues_.

In [45]:
Alberta_venues = getNearbyVenues(names=Alberta_data['Neighborhood'],
                                   latitudes=Alberta_data['Latitude'],
                                   longitudes=Alberta_data['Longitude']
                                  )

West Clareview, East Londonderry
North Capilano
East North Central, West Beverly
SE Capilano, West Southeast Industrial, East Bonnie Doon
Central Londonderry
Central Bonnie Doon
West Londonderry, East Calder
South Bonnie Doon, East University
North Central, Queen Mary Park, Blatchford
West University, Strathcona Place
North Downtown Fringe, East Downtown Fringe
Southgate, North Riverbend
North Downtown
Kaskitayo, Aspen Gardens
South Downtown, South Downtown Fringe (Alberta Provincial Government)
West Mill Woods
North Westmount, West Calder, East Mistatim
East Mill Woods
South Westmount, Groat Estate, East Northwest Industrial
Southwest Edmonton
Glenora, SW Downtown Fringe
South Industrial
North Jasper Place
East Southeast Industrial, South Clover Bar
Central Jasper Place, Buena Vista
Southgate, North Riverbend
West Northwest Industrial, Winterburn
North Clover Bar
West Jasper Place, West Edmonton Mall
The Meadows
Central Mistatim
The Palisades, West Castle Downs
Central Beverly
Heritag

#### Let's check the size of the resulting dataframe

In [46]:
print(Alberta_venues.shape)
Alberta_venues.head()

(320, 7)


Neighborhood Neighborhood Latitude  \
0  West Clareview, East Londonderry               53.5899   
1  West Clareview, East Londonderry               53.5899   
2  West Clareview, East Londonderry               53.5899   
3  West Clareview, East Londonderry               53.5899   
4  West Clareview, East Londonderry               53.5899   

  Neighborhood Longitude                     Venue  Venue Latitude  \
0              -113.4413     Buffet Royale Carvery       53.587229   
1              -113.4413              Café del Sol       53.592441   
2              -113.4413           Red Claw Gaming       53.586937   
3              -113.4413        My Grandma's Attic       53.586033   
4              -113.4413  Belvedere Transit Centre       53.587932   

   Venue Longitude      Venue Category  
0      -113.439075              Buffet  
1      -113.441455  Mexican Restaurant  
2      -113.439775    Toy / Game Store  
3      -113.441629         Record Shop  
4      -113.435254         Bus Station

Let's check how many venues were returned for each neighborhood.

In [47]:
Alberta_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Central Beverly                                                         4   
Central Bonnie Doon                                                     6   
Central Jasper Place, Buena Vista                                      10   
Central Mistatim                                                        3   
East Castledowns                                                        4   
East Mill Woods                                                         3   
East North Central, West Beverly                                        4   
East Southeast Industrial, South Clover Bar                             2   
Ellerslie                                                               4   
Glenora, SW Downtown Fringe                                             1   
Heritage Valley                                                         3   
Horse Hill, East Lake District                                          5   
Kaskitayo, Aspen Gardens                                                2   
North Capilano                                                          4   
North Central, Queen Mary Park, Blatchford                              7   
North Downtown                                                        100   
North Downtown Fringe, East Downtown Fringe                             6   
North Jasper Place                                                      4   
North Westmount, West Calder, East Mistatim                             4   
SE Capilano, West Southeast Industrial, East Bo...                      3   
South Bonnie Doon, East University                                      5   
South Downtown, South Downtown Fringe (Alberta ...                      9   
South Industrial                                                        3   
South Westmount, Groat Estate, East Northwest I...                      5   
Southgate, North Riverbend                                             14   
Southwest Edmonton                                                      2   
The Meadows                                                            20   
The Palisades, West Castle Downs                                        7   
West Clareview, East Londonderry                                        5   
West Jasper Place, West Edmonton Mall                                   9   
West Lake District                                                     18   
West Londonderry, East Calder                                           9   
West Mill Woods                                                         4   
West Northwest Industrial, Winterburn                                  21   
West University, Strathcona Place                                      10   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Central Beverly                                                          4   
Central Bonnie Doon                                                      6   
Central Jasper Place, Buena Vista                                       10   
Central Mistatim                                                         3   
East Castledowns                                                         4   
East Mill Woods                                                          3   
East North Central, West Beverly                                         4   
East Southeast Industrial, South Clover Bar                              2   
Ellerslie                                                                4   
Glenora, SW Downtown Fringe                                              1   
Heritage Valley                                                          3   
Horse Hill, East Lake District                                           5   
Kaskitayo, Aspen Gardens                                                 2   
North Capilano                  

Let's find out how many unique categories can be curated from all the returned venues

In [48]:
print('There are {} uniques categories.'.format(len(Alberta_venues['Venue Category'].unique())))

There are 129 uniques categories.


### Analyze Each Neighborhood

In [49]:
# one hot encoding
Alberta_onehot = pd.get_dummies(Alberta_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Alberta_onehot['Neighborhood'] = Alberta_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [Alberta_onehot.columns[-1]] + list(Alberta_onehot.columns[:-1])
Alberta_onehot = Alberta_onehot[fixed_columns]

Alberta_onehot.head()

Neighborhood  American Restaurant  Arts & Crafts Store  \
0  West Clareview, East Londonderry                    0                    0   
1  West Clareview, East Londonderry                    0                    0   
2  West Clareview, East Londonderry                    0                    0   
3  West Clareview, East Londonderry                    0                    0   
4  West Clareview, East Londonderry                    0                    0   

   Asian Restaurant  Auto Dealership  Bakery  Bank  Bar  Baseball Field  \
0                 0                0       0     0    0               0   
1                 0                0       0     0    0               0   
2                 0                0       0     0    0               0   
3                 0                0       0     0    0               0   
4                 0                0       0     0    0               0   

   Baseball Stadium  Big Box Store  Bookstore  Breakfast Spot  Brewery  \
0                 0              0          0               0        0   
1                 0              0          0               0        0   
2                 0              0          0               0        0   
3                 0              0          0               0        0   
4                 0              0          0               0        0   

   Buffet  Burger Joint  Bus Station  Business Service  Butcher  Café  Casino  \
0       1             0            0                 0        0     0       0   
1       0             0            0                 0        0     0       0   
2       0             0            0                 0        0     0       0   
3       0             0            0                 0        0     0       0   
4       0             0            1                 0        0     0       0   

   Cheese Shop  Chinese Restaurant  Clothing Store  Coffee Shop  College Gym  \
0            0                   0               0            0            0   
1            0                   0               0            0            0   
2            0                   0               0            0            0   
3            0                   0               0            0            0   
4            0                   0               0            0            0   

   College Residence Hall  Comic Shop  Construction & Landscaping  \
0                       0           0                           0   
1                       0           0                           0   
2                       0           0                           0   
3                       0           0                           0   
4                       0           0                           0   

   Convenience Store  Cosmetics Shop  Creperie  Department Store  \
0                  0               0         0                 0   
1                  0               0         0                 0   
2                  0               0         0                 0   
3                  0               0         0                 0   
4                  0               0         0                 0   

   Dim Sum Restaurant  Diner  Discount Store  Distribution Center  Dog Run  \
0                   0      0               0                    0        0   
1                   0      0               0                    0        0   
2                   0      0               0                    0        0   
3                   0      0               0                    0        0   
4                   0      0               0                    0        0   

   Eastern European Restaurant  Electronics Store  Fast Food Restaurant  \
0                            0                  0                     0   
1                            0                  0                     0   
2                            0                  0                     0   
3                            0                  0                     0   
4   

And let's examine the new dataframe size.

In [50]:
Alberta_onehot.shape

(320, 130)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [51]:
Alberta_grouped = Alberta_onehot.groupby('Neighborhood').mean().reset_index()
Alberta_grouped

Neighborhood  American Restaurant  \
0                                     Central Beverly             0.000000   
1                                 Central Bonnie Doon             0.166667   
2                   Central Jasper Place, Buena Vista             0.000000   
3                                    Central Mistatim             0.000000   
4                                    East Castledowns             0.000000   
5                                     East Mill Woods             0.000000   
6                    East North Central, West Beverly             0.000000   
7         East Southeast Industrial, South Clover Bar             0.000000   
8                                           Ellerslie             0.000000   
9                         Glenora, SW Downtown Fringe             0.000000   
10                                    Heritage Valley             0.000000   
11                     Horse Hill, East Lake District             0.000000   
12                           Kaskitayo, Aspen Gardens             0.000000   
13                                     North Capilano             0.000000   
14         North Central, Queen Mary Park, Blatchford             0.000000   
15                                     North Downtown             0.000000   
16        North Downtown Fringe, East Downtown Fringe             0.000000   
17                                 North Jasper Place             0.000000   
18        North Westmount, West Calder, East Mistatim             0.000000   
19  SE Capilano, West Southeast Industrial, East B...             0.000000   
20                 South Bonnie Doon, East University             0.200000   
21  South Downtown, South Downtown Fringe (Alberta...             0.000000   
22                                   South Industrial             0.000000   
23  South Westmount, Groat Estate, East Northwest ...             0.000000   
24                         Southgate, North Riverbend             0.000000   
25                                 Southwest Edmonton             0.000000   
26                                        The Meadows             0.000000   
27                   The Palisades, West Castle Downs             0.000000   
28                   West Clareview, East Londonderry             0.000000   
29              West Jasper Place, West Edmonton Mall             0.000000   
30                                 West Lake District             0.000000   
31                      West Londonderry, East Calder             0.000000   
32                                    West Mill Woods             0.000000   
33              West Northwest Industrial, Winterburn             0.000000   
34                  West University, Strathcona Place             0.000000   

    Arts & Crafts Store  Asian Restaurant  Auto Dealership    Bakery  \
0              0.000000              0.00             0.00  0.000000   
1              0.000000              0.00             0.00  0.000000   
2              0.000000              0.00             0.00  0.100000   
3              0.000000              0.00             0.00  0.000000   
4              0.000000              0.00             0.00  0.000000   
5              0.000000              0.00             0.00  0.333333   
6              0.000000              0.00             0.00  0.000000   
7              0.000000              0.00             0.00  0.000000   
8              0.000000              0.00             0.25  0.000000   
9              0.000000              0.00             0.00  0.000000   
10             0.000000              0.00             0.00  0.000000   
11             0.000000              0.00             0.00  0.000000   
12             0.000000              0.00             0.00  0.000000   
13             0.000000              0.00             0.00  0.000000   
14             0.000000              0.00             0.00  0.142857   
15             0.000000              0.02             0.00  0.000000   
16      

#### Let's confirm the new size

In [52]:
Alberta_grouped.shape

(35, 130)

Let's print each neighborhood along with the top 5 most common venues

In [53]:
num_top_venues = 5

for hood in Alberta_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = Alberta_grouped[Alberta_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Central Beverly----
                 venue  freq
0           Smoke Shop  0.50
1          IT Services  0.25
2        Grocery Store  0.25
3  American Restaurant  0.00
4               Office  0.00


----Central Bonnie Doon----
                 venue  freq
0  American Restaurant  0.17
1           Water Park  0.17
2       Cosmetics Shop  0.17
3                Trail  0.17
4         Liquor Store  0.17


----Central Jasper Place, Buena Vista----
                     venue  freq
0                     Café   0.1
1  Health & Beauty Service   0.1
2     Fast Food Restaurant   0.1
3        Convenience Store   0.1
4              Salad Place   0.1


----Central Mistatim----
                venue  freq
0        Liquor Store  0.33
1     Warehouse Store  0.33
2              Casino  0.33
3  Photography Studio  0.00
4         Pizza Place  0.00


----East Castledowns----
                venue  freq
0               Plaza  0.50
1          Playground  0.25
2        Skating Rink  0.25
3         Pizza Place 

Let's put that into a pandas dataframe

First, let's write a function to sort the venues in descending order.

In [54]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [55]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Alberta_grouped['Neighborhood']

for ind in np.arange(Alberta_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Alberta_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                    Central Beverly            Smoke Shop   
1                Central Bonnie Doon   American Restaurant   
2  Central Jasper Place, Buena Vista                  Café   
3                   Central Mistatim       Warehouse Store   
4                   East Castledowns                 Plaza   

  2nd Most Common Venue 3rd Most Common Venue    4th Most Common Venue  \
0           IT Services         Grocery Store      Filipino Restaurant   
1            Water Park        Cosmetics Shop                    Trail   
2           Salad Place           Pizza Place  Health & Beauty Service   
3                Casino          Liquor Store                Wine Shop   
4            Playground          Skating Rink        Electronics Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0   Fried Chicken Joint     French Restaurant            Food Truck   
1        Breakfast Spot          Liquor Store   Filipino Restaurant   
2                Bakery     Convenience Store        Sandwich Place   
3   Filipino Restaurant     French Restaurant            Food Truck   
4            Food Truck     Food & Drink Shop           Flower Shop   

  8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
0     Food & Drink Shop           Flower Shop         Fast Food Restaurant  
1     French Restaurant            Food Truck            Food & Drink Shop  
2  Fast Food Restaurant      Sushi Restaurant                 Liquor Store  
3     Food & Drink Shop           Flower Shop         Fast Food Restaurant  
4   Filipino Restaurant  Fast Food Restaurant  Eastern European Restaurant

### Cluster Neighborhoods

Run k-means to cluster the neighborhood into 5 clusters.

In [56]:
# set number of clusters
kclusters = 5

Alberta_grouped_clustering = Alberta_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Alberta_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 2, 2, 2, 1, 2, 2, 2, 2, 3], dtype=int32)

In [64]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Alberta_merged = Alberta_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
Alberta_merged = Alberta_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

ValueError: cannot insert Cluster Labels, already exists

In [66]:
Alberta_merged = Alberta_merged.dropna()

In [74]:
Alberta_merged= Alberta_merged.astype({"Cluster Labels": 'int64'})
Alberta_merged.head() # check the last columns!

Postal Code   Borough                                       Neighborhood  \
5          T5A  Edmonton                   West Clareview, East Londonderry   
6          T6A  Edmonton                                     North Capilano   
14         T5B  Edmonton                   East North Central, West Beverly   
15         T6B  Edmonton  SE Capilano, West Southeast Industrial, East B...   
24         T6C  Edmonton                                Central Bonnie Doon   

   Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
5   53.5899  -113.4413               2           Bus Station   
6   53.5483   -113.408               1            Playground   
14  53.5766  -113.4608               2            Smoke Shop   
15  53.5322  -113.4404               1            Playground   
24  53.5182  -113.4769               2   American Restaurant   

         2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
5             Toy / Game Store                Buffet           Record Shop   
6                         Park             Ski Trail           Bus Station   
14                 IT Services         Grocery Store   Filipino Restaurant   
15  Construction & Landscaping                   Bar  Fast Food Restaurant   
24                  Water Park        Cosmetics Shop                 Trail   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
5     Mexican Restaurant  Eastern European Restaurant     Electronics Store   
6         Discount Store          Distribution Center               Dog Run   
14   Fried Chicken Joint            French Restaurant            Food Truck   
15     French Restaurant                   Food Truck     Food & Drink Shop   
24        Breakfast Spot                 Liquor Store   Filipino Restaurant   

          8th Most Common Venue 9th Most Common Venue  \
5                       Dog Run   Distribution Center   
6   Eastern European Restaurant                 Diner   
14            Food & Drink Shop           Flower Shop   
15                  Flower Shop   Filipino Restaurant   
24            French Restaurant            Food Truck   

         10th Most Common Venue  
5                   Gas Station  
6          Fast Food Restaurant  
14         Fast Food Restaurant  
15  Eastern European Restaurant  
24            Food & Drink Shop

In [75]:
Alberta_merged

Postal Code   Borough                                       Neighborhood  \
5           T5A  Edmonton                   West Clareview, East Londonderry   
6           T6A  Edmonton                                     North Capilano   
14          T5B  Edmonton                   East North Central, West Beverly   
15          T6B  Edmonton  SE Capilano, West Southeast Industrial, East B...   
24          T6C  Edmonton                                Central Bonnie Doon   
32          T5E  Edmonton                      West Londonderry, East Calder   
33          T6E  Edmonton                 South Bonnie Doon, East University   
41          T5G  Edmonton         North Central, Queen Mary Park, Blatchford   
42          T6G  Edmonton                  West University, Strathcona Place   
50          T5H  Edmonton        North Downtown Fringe, East Downtown Fringe   
51          T6H  Edmonton                         Southgate, North Riverbend   
59          T5J  Edmonton                                     North Downtown   
60          T6J  Edmonton                           Kaskitayo, Aspen Gardens   
68          T5K  Edmonton  South Downtown, South Downtown Fringe (Alberta...   
69          T6K  Edmonton                                    West Mill Woods   
77          T5L  Edmonton        North Westmount, West Calder, East Mistatim   
78          T6L  Edmonton                                    East Mill Woods   
86          T5M  Edmonton  South Westmount, Groat Estate, East Northwest ...   
87          T6M  Edmonton                                 Southwest Edmonton   
95          T5N  Edmonton                        Glenora, SW Downtown Fringe   
96          T6N  Edmonton                                   South Industrial   
104         T5P  Edmonton                                 North Jasper Place   
105         T6P  Edmonton        East Southeast Industrial, South Clover Bar   
113         T5R  Edmonton                  Central Jasper Place, Buena Vista   
114         T6R  Edmonton                         Southgate, North Riverbend   
122         T5S  Edmonton              West Northwest Industrial, Winterburn   
131         T5T  Edmonton              West Jasper Place, West Edmonton Mall   
132         T6T  Edmonton                                        The Meadows   
140         T5V  Edmonton                                   Central Mistatim   
141         T6V  Edmonton                   The Palisades, West Castle Downs   
149         T5W  Edmonton                                    Central Beverly   
150         T6W  Edmonton                                    Heritage Valley   
158         T5X  Edmonton                                   East Castledowns   
159         T6X  Edmonton                                          Ellerslie   
167         T5Y  Edmonton                     Horse Hill, East Lake District   
176         T5Z  Edmonton                                 West Lake District   

    Latitude  Longitude  Cluster Labels      1st Most Common Venue  \
5    53.5899  -113.4413               2                Bus Station   
6    53.5483   -113.408               1                 Playground   
14   53.5766  -113.4608               2                 Smoke Shop   
15   53.5322  -113.4404               1                 Playground   
24   53.5182  -113.4769               2        American Restaurant   
32   53.5923  -113.5168               2              Grocery Store   
33   53.5087  -113.5078               2        American Restaurant   
41   53.5682  -113.4822               2                   Pharmacy   
42   53.5248  -113.5334               2                    Theater   
50    53.555  -113.4822               2            Thai Restaurant   
51   53.4839  -113.5227               2     Furniture / Home Store   
59   53.5421  -113.4989               2                Coffee Shop   
60   53.4822  -113.5269               0                       Lake   
68    53.535   -113.501               1                       P

Finally, let's visualize the resulting clusters

In [81]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Alberta_merged['Latitude'], Alberta_merged['Longitude'], Alberta_merged['Neighborhood'], Alberta_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster],
        fill=True,
        fill_color=rainbow[cluster],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examine Clusters

Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster.

#### Cluster 0

In [88]:
Alberta_merged.loc[Alberta_merged['Cluster Labels'] == 0, Alberta_merged.columns[[1] + list(range(5, Alberta_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
60  Edmonton               0                  Lake          Tennis Court   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
60             Wine Shop  Fast Food Restaurant     French Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
60            Food Truck     Food & Drink Shop           Flower Shop   

   9th Most Common Venue 10th Most Common Venue  
60   Filipino Restaurant      Electronics Store

#### Cluster 1

In [84]:
Alberta_merged.loc[Alberta_merged['Cluster Labels'] == 1, Alberta_merged.columns[[1] + list(range(5, Alberta_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
6    Edmonton               1            Playground   
15   Edmonton               1            Playground   
68   Edmonton               1                  Park   
96   Edmonton               1                  Park   
158  Edmonton               1                 Plaza   

          2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
6                          Park             Ski Trail           Bus Station   
15   Construction & Landscaping                   Bar  Fast Food Restaurant   
68              Thai Restaurant                 Hotel                 Plaza   
96           Golf Driving Range                   Gym             Wine Shop   
158                  Playground          Skating Rink     Electronics Store   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
6          Discount Store   Distribution Center               Dog Run   
15      French Restaurant            Food Truck     Food & Drink Shop   
68      Convenience Store        Sandwich Place     French Restaurant   
96    Filipino Restaurant     French Restaurant            Food Truck   
158            Food Truck     Food & Drink Shop           Flower Shop   

           8th Most Common Venue 9th Most Common Venue  \
6    Eastern European Restaurant                 Diner   
15                   Flower Shop   Filipino Restaurant   
68              Baseball Stadium  Fast Food Restaurant   
96             Food & Drink Shop           Flower Shop   
158          Filipino Restaurant  Fast Food Restaurant   

          10th Most Common Venue  
6           Fast Food Restaurant  
15   Eastern European Restaurant  
68             Food & Drink Shop  
96          Fast Food Restaurant  
158  Eastern European Restaurant

#### Cluster 2

In [85]:
Alberta_merged.loc[Alberta_merged['Cluster Labels'] == 2, Alberta_merged.columns[[1] + list(range(5, Alberta_merged.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
5    Edmonton               2                Bus Station   
14   Edmonton               2                 Smoke Shop   
24   Edmonton               2        American Restaurant   
32   Edmonton               2              Grocery Store   
33   Edmonton               2        American Restaurant   
41   Edmonton               2                   Pharmacy   
42   Edmonton               2                    Theater   
50   Edmonton               2            Thai Restaurant   
51   Edmonton               2     Furniture / Home Store   
59   Edmonton               2                Coffee Shop   
69   Edmonton               2           Business Service   
77   Edmonton               2  Middle Eastern Restaurant   
78   Edmonton               2                        Pub   
86   Edmonton               2       Fast Food Restaurant   
104  Edmonton               2        Rental Car Location   
105  Edmonton               2        Housing Development   
113  Edmonton               2                       Café   
114  Edmonton               2     Furniture / Home Store   
122  Edmonton               2                      Hotel   
131  Edmonton               2              Hot Dog Joint   
132  Edmonton               2                 Restaurant   
140  Edmonton               2            Warehouse Store   
141  Edmonton               2             Ice Cream Shop   
149  Edmonton               2                 Smoke Shop   
150  Edmonton               2             Gymnastics Gym   
159  Edmonton               2       Fast Food Restaurant   
167  Edmonton               2                   Pharmacy   
176  Edmonton               2                Coffee Shop   

          2nd Most Common Venue  3rd Most Common Venue  \
5              Toy / Game Store                 Buffet   
14                  IT Services          Grocery Store   
24                   Water Park         Cosmetics Shop   
32          Arts & Crafts Store                Dog Run   
33                  Coffee Shop               Pharmacy   
41                       Bakery                   Bank   
42                  College Gym                    Pub   
50                          Gym         Soccer Stadium   
51                  Coffee Shop         Sandwich Place   
59               Sandwich Place                    Pub   
69         Gym / Fitness Center           Home Service   
77                          Pub         Breakfast Spot   
78                       Bakery         Massage Studio   
86                  Pizza Place                   Bank   
104                Home Service      Convenience Store   
105  Construction & Landscaping     Dim Sum Restaurant   
113                 Salad Place            Pizza Place   
114                 Coffee Shop         Sandwich Place   
122        Fast Food Restaurant  Vietnamese Restaurant   
131        Medical Supply Store               Pharmacy   
132        Fast Food Restaurant            Coffee Shop   
140                      Casino           Liquor Store   
141                  Restaurant               Pharmacy   
149                 IT Services          Grocery Store   
150                   Rest Area   Fast Food Restaurant   
159             Auto Dealership        Motorcycle Shop   
167            Department Store            Coffee Shop   
176        Fast Food Restaurant               Pharmacy   

             4th Most Common Venue   5th Most Common Venue  \
5                      Record Shop      Mexican Restaurant   
14             Filipino Restaurant     Fried Chicken Joint   
24                           Trail          Breakfast Spot   
32                         Butcher                  Bakery   
33        Mediterranean Restaurant             Flower Shop   
41                     Music Venue                 Theater   
42                  Sandwich Place             Coffee Shop   
50                       Gift Shop           Grocery Store   
51                      

#### Cluster 3

In [86]:
Alberta_merged.loc[Alberta_merged['Cluster Labels'] == 3, Alberta_merged.columns[[1] + list(range(5, Alberta_merged.shape[1]))]]

Borough  Cluster Labels  1st Most Common Venue 2nd Most Common Venue  \
95  Edmonton               3  Portuguese Restaurant             Wine Shop   

     3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
95  Furniture / Home Store                 Diner        Discount Store   

   6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
95   Distribution Center               Dog Run  Eastern European Restaurant   

   9th Most Common Venue 10th Most Common Venue  
95     Electronics Store   Fast Food Restaurant

#### Cluster 4

In [87]:
Alberta_merged.loc[Alberta_merged['Cluster Labels'] == 4, Alberta_merged.columns[[1] + list(range(5, Alberta_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
87  Edmonton               4       Paintball Field             Wine Shop   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
87    Dim Sum Restaurant        Discount Store   Distribution Center   

   6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
87               Dog Run  Eastern European Restaurant     Electronics Store   

   9th Most Common Venue 10th Most Common Venue  
87  Fast Food Restaurant    Filipino Restaurant

## Result <a name="results"></a>

Finally, we can find the good location for opening a restaurant in **Edmonton City, Alberta**, Canada.Edmonton City has 1 boroughs and 38 neighborhoods.
We try to detect **locations that are not already crowded with restaurants**. We would also prefer locations **as close to city center as possible**.
From Edmonton's neighborhood, we found 7 venues, 129 uniques categories. were returned by Foursquare within a radius of 500 meters.
And we cluster the Edmonton's neighborhood to 5 clusters by using k-means. 
The result from clustering, Cluster no. 1,2 have many restuarants and also have Thai restuarants in most common venue.
But Cluster no. 0,3,4 don't have Thai restuarants. We will consider cluster no. 0,3,4.

## Discussion <a name="discussion"></a>
For Cluster no. 0,3,4, don't have Thai restuarants. I prefer to select cluster no.3 because this zone is close to city center.
But cluster no.4 have many Asian restuarants, we need to seperate from the exising Asian restuarants. And cluster no.0 is far from city center more than cluster no.3

## Conclusion <a name="conclusion"></a>

Purpose of this project was to identify Edmonton areas close to Alberta center. Clustering of those locations was then performed in order to create zones of interest and addresses of those zone centers were created to be used as starting points for final exploration by stakeholders.

Final decision about optimal restaurant location will be made by stakeholders based on specific characteristics of neighborhoods and locations in every recommended zone, taking into consideration additional factors like attractiveness of each location (proximity to park or water), levels of noise / proximity to major roads, real estate availability, prices, social and economic dynamics of every neighborhood etc.